File: java/com/google/copybara/git/gerritapi/ProjectInfo.java
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git.gerritapi;

import com.google.api.client.util.Key;
import com.google.common.base.MoreObjects;
import java.util.Map;
import net.starlark.java.eval.Printer;
import net.starlark.java.eval.StarlarkValue;

/** See https://gerrit-review.googlesource.com/Documentation/rest-api-projects.html#project-info */
public class ProjectInfo implements StarlarkValue {
  @Key String id;
  @Key String name;
  @Key String parent;
  @Key String description;
  @Key String state;
  @Key Map<String, String> branches;

  public enum State {
    ACTIVE, READ_ONLY, HIDDEN
  }

  public String getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public String getParent() {
    return parent;
  }

  public String getDescription() {
    return description;
  }

  public State getState() {
    return state == null ? null : State.valueOf(state);
  }

  public Map<String, String> getBranches() {
    return branches;
  }

  @Override
  public void repr(Printer printer) {
    printer.append(toString());
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper(this)
        .add("id", id)
        .add("name", name)
        .add("parent", parent)
        .add("description", description)
        .add("state", state)
        .add("branches", branches)
        .toString();
  }
}

File: src/Copybara.Core/Git/GerritApi/ProjectInfo.cs
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Text.Json.Serialization;
using Starlark.Eval;

namespace Copybara.Git.GerritApi;

/// <summary>See https://gerrit-review.googlesource.com/Documentation/rest-api-projects.html#project-info</summary>
public class ProjectInfo : IStarlarkPrintableValue
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("parent")]
    public string? Parent { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("state")]
    public string? StateString { get; set; }

    [JsonPropertyName("branches")]
    public IReadOnlyDictionary<string, string>? Branches { get; set; }

    public enum State
    {
        ACTIVE,
        READ_ONLY,
        HIDDEN,
    }

    public string? GetId() => Id;

    public string? GetName() => Name;

    public string? GetParent() => Parent;

    public string? GetDescription() => Description;

    public State? GetState() => StateString is null ? null : Enum.Parse<State>(StateString);

    public IReadOnlyDictionary<string, string>? GetBranches() => Branches;

    public void Repr(Printer printer, StarlarkSemantics semantics) => printer.Append(ToString());

    public override string ToString() =>
        $"ProjectInfo{{id={Id}, name={Name}, parent={Parent}, description={Description}, "
        + $"state={StateString}, branches={Branches}}}";
}
